### PR TITLE
Add merge support and unit tests

### PR DESCRIPTION
**Pull request by Zyiqin-Miranda**
## Description:
This PR adds initial merge support and unit tests for merge. 
## Testing:
Unit tests.
## This PR includes:
- [x] Skipping manifests for hash buckets are not touched.
- [x] Empty primary key list.
- [x] Delete type deltas, empty hash bucket after applying delete. 
- [x] Create concatenated string in hash bucketed table in merge for deduping.
- [x] Bug in drop_duplicates not being deterministic. 
- [x] Deduped record count in merge log and assertion not calculated correctly right now.
- [x] Old compacted table can have empty hash bucket
- [x] Do not drop duplicates during rebase.

### CODE REVIEW

File: deltacat/compute/compactor_v2/steps/merge.py [hunk 1]
```diff
@@ -1,18 +1,439 @@
...
+                    f"[Merge task index {input.merge_task_index}] Merged "
+                    f"record count: {len(table)}, took: {merge_time}s"
+                )
+
+                materialized_results.append(_materialize(hb_idx, [table]))
+            else:
+                hb_index_copy_by_reference.append(hb_idx)
+
+        if input.round_completion_info and hb_index_copy_by_reference:
+            referenced_materialized_results = (
```
Comment (rkenmi) [inline deltacat/compute/compactor_v2/steps/merge.py:412]: Can we add logging during reference copy as well?

File: deltacat/compute/compactor_v2/steps/merge.py [hunk 2]
```diff
@@ -1,18 +1,439 @@
...
+
+def _download_compacted_table(
+    hb_index: int,
+    rcf: RoundCompletionInfo,
+    read_kwargs_provider: Optional[ReadKwargsProvider] = None,
+    deltacat_storage=unimplemented_deltacat_storage,
+    deltacat_storage_kwargs: Optional[dict] = None,
+) -> pa.Table:
+    tables = []
+    hb_index_to_indices = rcf.hb_index_to_entry_range
```
Comment (rkenmi) [inline deltacat/compute/compactor_v2/steps/merge.py:184]: A RCF may not necessarily have`hb_index_to_entry_range` right?

File: deltacat/compute/compactor_v2/steps/merge.py [hunk 3]
```diff
@@ -1,18 +1,439 @@
...
+            f"Unexpected Error: Materialized delta manifest record count "
+            f"({manifest_records}) does not equal compacted table record count "
+            f"({len(compacted_table)})"
+        )
+        materialize_result = MaterializeResult.of(
+            delta=delta,
+            task_index=hash_bucket_index,
+            # TODO (pdames): Generalize WriteResult to contain in-memory-table-type
+            #  and in-memory-table-bytes instead of tight coupling to paBytes
+            pyarrow_write_result=PyArrowWriteResult.of(
```
Comment (rkenmi) [inline deltacat/compute/compactor_v2/steps/merge.py:306]: Should we also capture the Parquet format version somewhere during Parquet writes?

File: deltacat/compute/compactor/model/delta_annotated.py [hunk 4]
```diff
@@ -89,6 +89,11 @@ def rebatch(
         for delta_annotated in annotated_deltas:
             split_annotated_deltas.extend(DeltaAnnotated._split_single(delta_annotated))
```
[the remark this answers is not in the document]
Comment (raghumdani) [reply deltacat/compute/compactor/model/delta_annotated.py:90]: Yeah, there is a 10s of seconds of added latency atleast. Hence, this is only enabled in incremental where we saw the error. Rebase files would be uniform anyways.

File: deltacat/compute/compactor_v2/compaction_session.py [hunk 5]
```diff
@@ -182,7 +180,22 @@ def _execute_compaction(
...
+        min_delta_bytes=params.min_delta_bytes_in_batch,
+        min_file_counts=params.min_files_in_batch,
+        # disable input split during rebase as the rebase files are already uniform
+        enable_input_split=params.rebase_source_partition_locator is None,
+        deltacat_storage_kwargs=params.deltacat_storage_kwargs,
+    )
+
     delta_discovery_end = time.monotonic()
+
+    compaction_audit.set_uniform_deltas_created(len(uniform_deltas))
```
[the remark this answers is not in the document]
Comment (raghumdani) [reply deltacat/compute/compactor_v2/compaction_session.py:198]: Sure.

File: deltacat/compute/compactor_v2/steps/merge.py [hunk 6]
```diff
@@ -1,18 +1,435 @@
...
+        key=lambda df: (df.stream_position, df.file_index),
+        reverse=False,  # ascending
+    )
+    is_delete = False
+    for df_envelope in df_envelopes:
+        assert (
+            df_envelope.delta_type != DeltaType.APPEND
+        ), "APPEND type deltas are not supported. Kindly use UPSERT or DELETE"
+        if df_envelope.delta_type == DeltaType.DELETE:
+            is_delete = True
```
[the remark this answers is not in the document]
Comment (raghumdani) [reply deltacat/compute/compactor_v2/steps/merge.py:98]: 1. There can only be two values UPSERT and DELETE (previous compactor supported only those). 
2. Yes, there can be any order of dfes like UPSERT, and then DELETE and then UPSERT.

File: deltacat/compute/compactor_v2/steps/merge.py [hunk 7]
```diff
@@ -1,18 +1,435 @@
...
+    all_tables = []
+    incremental_idx = 0
+
+    if compacted_table:
+        incremental_idx = 1
+        all_tables.append(compacted_table)
+
+    all_tables.append(table)
+
+    if not primary_keys or not can_drop_duplicates:
```
[the remark this answers is not in the document]
Comment (raghumdani) [reply deltacat/compute/compactor_v2/steps/merge.py:136]: Sure..

File: deltacat/compute/compactor_v2/steps/merge.py (hunk 2 above)
Comment (raghumdani) [reply deltacat/compute/compactor_v2/steps/merge.py:184]: If you are downloading compacted table, there has to be rcf. See the code below that checks this.

File: deltacat/compute/compactor_v2/steps/merge.py [hunk 8]
```diff
@@ -1,18 +1,435 @@
...
+        compacted_delta_locator, **deltacat_storage_kwargs
+    )
+
+    manifest_entry_referenced_list = []
+    materialize_result_list = []
+    hb_index_to_indices = round_completion_info.hb_index_to_entry_range
+
+    for hb_index in hb_index_copy_by_reference:
+        if str(hb_index) not in hb_index_to_indices:
+            continue
```
[the remark this answers is not in the document]
Comment (raghumdani) [reply deltacat/compute/compactor_v2/steps/merge.py:227]: When the hash bucket is not touched both in compacted table and incremental.

File: deltacat/compute/compactor_v2/steps/merge.py [hunk 9]
```diff
@@ -1,18 +1,435 @@
...
+            delta_type=DeltaType.UPSERT,
+            meta=manifest.meta,
+            manifest=manifest,
+            previous_stream_position=write_to_partition.stream_position,
+            properties={},
+        )
+        materialize_result = MaterializeResult.of(
+            delta=delta,
+            task_index=hb_index,
+            pyarrow_write_result=None,
```
[the remark this answers is not in the document]
Comment (raghumdani) [reply deltacat/compute/compactor_v2/steps/merge.py:252]: Yes, we separated that out earlier. Using the same mechanism.

File: deltacat/compute/compactor_v2/steps/merge.py (hunk 3 above)
Comment (raghumdani) [reply deltacat/compute/compactor_v2/steps/merge.py:306]: parquet format version is currently handled as kwargs. We should already be auditing the kwargs.

File: deltacat/compute/compactor_v2/steps/merge.py (hunk 1 above)
Comment (raghumdani) [reply deltacat/compute/compactor_v2/steps/merge.py:412]: Sure.

File: deltacat/tests/compute/compactor_v2/steps/test_merge.py [hunk 10]
```diff
@@ -0,0 +1,423 @@
...
+                    object_store=object_store,
+                )
+            )
+        merge_res_list = []
+        for merge_input in merge_input_list:
+            merge_result_promise = merge.remote(merge_input)
+            merge_result = ray.get(merge_result_promise)
+            merge_res_list.append(merge_result)
+
+        # All records vanish
```
[the remark this answers is not in the document]
Comment (raghumdani) [reply deltacat/tests/compute/compactor_v2/steps/test_merge.py:295]: Pierce is adding that as part of this CR.

File: deltacat/compute/compactor_v2/steps/merge.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (rkenmi) [reply deltacat/compute/compactor_v2/steps/merge.py:98]: I'm curious about the impact of severe skews here with huge tables of UPSERT type and, say only a few small DELETE type tables. It looks like we append a delete bool PyArrow table column for every table as long as we encounter at least one DELETE type delta.

File: deltacat/compute/compactor_v2/steps/merge.py (hunk 2 above)
Comment (rkenmi) [reply deltacat/compute/compactor_v2/steps/merge.py:184]: We might also need a check for `_copy_all_manifest_files_from_old_hash_buckets` as well if we want backwards compatibility with old RCFs that might not have this value.

File: deltacat/compute/compactor_v2/steps/merge.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (raghumdani) [reply deltacat/compute/compactor_v2/steps/merge.py:98]: In most cases, we will only find UPSERT, so we don't append the delta_type column. In other cases, we will only find DELETE as compaction runs on every delta.

File: deltacat/compute/compactor_v2/steps/merge.py (hunk 2 above)
Comment (raghumdani) [reply deltacat/compute/compactor_v2/steps/merge.py:184]: What I meant is, we call this function only if hb_index_to_entry_range is present. 

```python
if ( input.round_completion_info
                    and input.round_completion_info.hb_index_to_entry_range
                    and input.round_completion_info.hb_index_to_entry_range.get(
                        str(hb_idx)
                    )
                    is not None
                )
```

Comment (rkenmi) [reply deltacat/compute/compactor_v2/steps/merge.py:184]: For downloading compacted tables, that makes sense. Are we sure we're performing the same kind of check when copying manifests though?

Comment (raghumdani) [reply deltacat/compute/compactor_v2/steps/merge.py:184]: Got it.. Addressed.

**Review by rkenmi — APPROVED** [review]

**Pull request merged**